server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java

```diff
@@ -94,24 +107,59 @@ protected SearchPhase getNextPhase(SearchPhaseResults<SearchService.CanMatchResp
                 iter.resetAndSkip();
             }
         }
-        return shardsIts;
+        if (shouldSortShards(results.sortValues) == false) {
+            return shardsIts;
+        }
+        int sortMul = FieldSortBuilder.getPrimaryFieldSortOrNull(source).order() == SortOrder.ASC ? 1 : -1;
```

Comment:
I know Lucene does this, but can we do something cleaner that doesn't involve multiplying the output of comparators (since Integer#MIN_VALUE is a legal return value) here, e.g. (pseudo code)
```java
Comparator comparator;
if (order == ASC) {
  comparator = = Comparator.naturalOrder();
} else {
  comparator = Comparator.reverseOrder();
}
```
and passing this comparator instead of `sortMul`?

Reply:
+1, I pushed https://github.com/elastic/elasticsearch/pull/49092/commits/9f959dc1db43577a02c1f5b9864c211e235db5e2 to use `Comparator#reversed` in this case.